app/maptools/recordingmaptool.cpp

```diff
@@ -568,19 +566,69 @@ void RecordingMapTool::releaseVertex( const QgsPoint &point )
     return;
   }
 
-  if ( mRecordedGeometry.type() == QgsWkbTypes::PolygonGeometry && mRecordedGeometry.constGet()->nCoordinates() == 2 )
-  {
-    setState( MapToolState::Record );
-  }
-  else if ( mRecordedGeometry.constGet()->nCoordinates() == 1 )
+  int vertexCount = mRecordedGeometry.constGet()->vertexCount( mActiveVertex.vertexId().part, mActiveVertex.vertexId().ring );
+
+  if ( mRecordedGeometry.type() == QgsWkbTypes::PolygonGeometry )
   {
-    updateVertex( mActiveVertex, point );
-    setState( MapToolState::Record );
+    QgsLineString *r;
+
+    if ( mActiveVertex.vertexId().ring == 0 )
+    {
+      r = qgsgeometry_cast<QgsLineString *>( qgsgeometry_cast<const QgsPolygon *>( mRecordedGeometry.constGet() )->exteriorRing() );
+    }
+    else
+    {
+      r = qgsgeometry_cast<QgsLineString *>( qgsgeometry_cast<const QgsPolygon *>( mRecordedGeometry.constGet() )->interiorRing( mActiveVertex.vertexId().ring ) );
+    }
+
+    if ( !r )
+    {
+      return;
+    }
+
+    if ( vertexCount == 2 )
+    {
+      r->close();
+      emit recordedGeometryChanged( mRecordedGeometry );
+
+      setState( MapToolState::Record );
+      setActiveVertex( Vertex() );
+      return;
+    }
+    else if ( vertexCount == 1 )
+    {
+      updateVertex( mActiveVertex, point );
+      setState( MapToolState::Record );
+      setActiveVertex( Vertex() );
+      return;
+    }
   }
-  else
+
+  updateVertex( mActiveVertex, point );
+
+  // if it is a first or last vertex of the line we go to the recording mode
+  if ( mRecordedGeometry.type() == QgsWkbTypes::LineGeometry )
   {
-    setState( MapToolState::View );
+    if ( mActiveVertex.type() != Vertex::Existing && mActiveVertex.vertexId().vertex == 0 )
```

Comment:
Don't we actually want it to be existing? (and only existing?) 
```suggestion
    if ( mActiveVertex.type() == Vertex::Existing && mActiveVertex.vertexId().vertex == 0 )
```

Reply:
In this case releasing start/end point after grabbing it will end up in recording state, no? I thought we want go to recording only after removing handle, so grabbing start/end will be used only to move it in the new place.

Reply:
> With this change we will always enter recording mode when start/end vertex is released.

Yes, that is correct and expected 🙂